insight_testsuite/temp/src/lib.py

```diff
@@ -0,0 +1,169 @@
+    ]
+    return session_list
+
+
+class ActiveSessions(OrderedDict):
+    """
+    Core data structure for EDGAR sessionization.
+
+    Records the evolving set of active sessions, clearing
+    inactive sessions as new requests come in.
+    """
+
+    def __init__(self, inactivity_period):
+        """
+        Keyword Arguments:
+            inactivity_period {int} -- Number of seconds a session
+            can remain active without requests
+        """
+        OrderedDict.__init__(self)
+        self.inactivity_period = inactivity_period
+
+    def __missing__(self, key):
+        """
+        Implements DefaultDict-like functionality to automatically return
+        a new session dict object when an ip for a request is not currently
+        active
+        """
+        return {
+            "first_request": self.current_datetime,
+            "last_request": self.current_datetime,
+            "request_count": 0,
+            "ip": key
+        }
+
+    def _isinactive(self, session):
+        seconds_old = seconds_delta(
+            self.current_datetime, session["last_request"])
+        return seconds_old > self.inactivity_period
+
+    def _update_session(self, processed_row_dict):
+        """Update active session's 'count' and 'last_request'.
+        If session does not exist that matches 'ip' creates a new one.
+
+        Private method called by self.step
+
+        Arguments:
+            processed_row_dict {dict} -- corresponds to single row
+            from input log.csv, after pre-processing by process_row function
+        """
+        session = self[processed_row_dict["ip"]].copy()
+        session["last_request"] = processed_row_dict["datetime"]
+        session["request_count"] += 1
+        self[processed_row_dict["ip"]] = session
+
+    def _close_sessions(self, close_all=False):
+        """
+        Close inactive sessions, if close_all=True then close all
+        sessions.
+
+        Private method called by self.step and self.final_step
+        """
+        if close_all:
+            self.closed_sessions = list(self.values())
+        else:
+            self.closed_sessions = []
+            for _, session in self.items():
+                if self._isinactive(session):
+                    self.closed_sessions.append(session)
+
+        # Clear closed sessions
+        for closed_session in self.closed_sessions:
+            closed_ip = closed_session["ip"]
+            del self[closed_ip]
+
+    def step(self, processed_row_dict):
+        """Evolve state by one timestep: update an active session /
+        create a new one. Close inactive sessions.
+
+        Arguments:
+            processed_row_dict {dict} -- corresponds to single row
+            from input log.csv, after pre-processing by process_row function
+
+        Returns:
+            [list] -- List of closed session dictionaries
+        """
+        self.current_datetime = processed_row_dict["datetime"]
+        self._close_sessions()
```

Comment:
So you are checking every line if any of the sessions need to be closed. It will be a bit faster if you skip this when no time has elapsed.